"is not an iterator" diagnostic through blanket implementation, when the type is actually an iterator but is missing a marker

### Code

(inspired by [rayon's `ParallelBridge`](https://docs.rs/rayon/latest/rayon/iter/trait.ParallelBridge.html))

```Rust
pub trait ParallelBridge: Sized {
    // Required method
    fn par_bridge(self) -> ();
}

impl<T: Iterator + Send> ParallelBridge for T
where
    T::Item: Send,
{
    fn par_bridge(self) -> () {
        unimplemented!("par_bridge");
    }
}


fn f(it: impl Iterator<Item=u64>) {
    it.map(|i| i+1).par_bridge()
}

fn main() -> std::io::Result<()> {
    let (_tx, rx) = std::sync::mpsc::sync_channel(1000);
    f(rx.iter());

    Ok(())
}
```

[playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=290bd6102ce2b63a30bb6c78e5a74e8b)

### Current output

```Shell
Compiling playground v0.0.1 (/playground)
error[E0599]: no method named `par_bridge` found for struct `Map<impl Iterator<Item = u64>, {closure@src/main.rs:17:12: 17:15}>` in the current scope
  --> src/main.rs:17:21
   |
17 |     it.map(|i| i+1).par_bridge()
   |                     ^^^^^^^^^^ `Map<impl Iterator<Item = u64>, {closure@src/main.rs:17:12: 17:15}>` is not an iterator
   |
help: call `.into_iter()` first
   |
17 |     it.map(|i| i+1).into_iter().par_bridge()
   |                     ++++++++++++

For more information about this error, try `rustc --explain E0599`.
error: could not compile `playground` (bin "playground") due to 1 previous error
```

### Desired output

The same error there is when removing the `.map()`

```
   Compiling playground v0.0.1 (/playground)
error[E0599]: no method named `par_bridge` found for type parameter `impl Iterator<Item = u64>` in the current scope
  --> src/main.rs:17:8
   |
16 | fn f(it: impl Iterator<Item=u64>) {
   |          ----------------------- method `par_bridge` not found for this type parameter
17 |     it.map(|i| i+1).par_bridge()
   |        ^^^^^^^^^^ method cannot be called on `impl Iterator<Item = u64>` due to unsatisfied trait bounds
   |
   = help: items from traits can only be used if the type parameter is bounded by the trait
help: the following trait defines an item `par_bridge`, perhaps you need to restrict type parameter `impl Iterator<Item = u64>` with it:
   |
16 | fn f(it: impl Iterator<Item=u64> + ParallelBridge) {
   |                                  ++++++++++++++++

For more information about this error, try `rustc --explain E0599`.
error: could not compile `playground` (bin "playground") due to 1 previous error
```


### Rationale and extra context

The error indicated the type does not implement `Iterator`, even though it does and `Send` is the one that is missing.

### Other cases

_No response_

### Rust Version

```Shell
rustc 1.79.0
```


### Anything else?

_No response_

Here is a slightly smaller reproduction that only uses `core` and has a more egregious error message (“`impl Iterator` is not an iterator”). Reduced from a case found by @zakarumych.

```rust
trait Missing {}
trait HasMethod {
    fn foo(self);
}
impl<T: Iterator + Missing> HasMethod for T {
    fn foo(self) {}
}

fn get_iter() -> impl Iterator {
    // any Iterator impl will do
    core::iter::once(())
}

fn main() {
    get_iter().foo();
}
```

```
error[E0599]: no method named `foo` found for opaque type `impl Iterator` in the current scope
  --> src/main.rs:15:16
   |
15 |     get_iter().foo();
   |                ^^^ `impl Iterator` is not an iterator
   |
help: call `.into_iter()` first
   |
15 |     get_iter().into_iter().foo();
   |                ++++++++++++
```

--- a/compiler/rustc_hir_typeck/src/method/suggest.rs
+++ b/compiler/rustc_hir_typeck/src/method/suggest.rs
@@ -100,9 +100,9 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         ty: Ty<'tcx>,
         span: Span,
         unsatisfied_predicates: &Vec<(
-            ty::Predicate<'_>,
-            Option<ty::Predicate<'_>>,
-            Option<ObligationCause<'_>>,
+            ty::Predicate<'tcx>,
+            Option<ty::Predicate<'tcx>>,
+            Option<ObligationCause<'tcx>>,
         )>,
     ) -> bool {
         fn predicate_bounds_generic_param<'tcx>(
@@ -131,15 +131,17 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             }
         }
 
-        fn is_iterator_predicate(predicate: ty::Predicate<'_>, tcx: TyCtxt<'_>) -> bool {
+        let is_iterator_predicate = |predicate: ty::Predicate<'tcx>| -> bool {
             if let ty::PredicateKind::Clause(ty::ClauseKind::Trait(trait_pred)) =
                 predicate.kind().as_ref().skip_binder()
             {
-                tcx.is_diagnostic_item(sym::Iterator, trait_pred.trait_ref.def_id)
+                self.tcx.is_diagnostic_item(sym::Iterator, trait_pred.trait_ref.def_id)
+                    // ignore unsatisfied predicates generated from trying to auto-ref ty (#127511)
+                    && trait_pred.trait_ref.self_ty() == ty
             } else {
                 false
             }
-        }
+        };
 
         // Does the `ty` implement `IntoIterator`?
         let Some(into_iterator_trait) = self.tcx.get_diagnostic_item(sym::IntoIterator) else {
@@ -164,7 +166,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                         generics,
                         generic_param,
                         self.tcx,
-                    ) && is_iterator_predicate(unsatisfied.0, self.tcx)
+                    ) && is_iterator_predicate(unsatisfied.0)
                     {
                         return true;
                     }
@@ -172,7 +174,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             }
             ty::Slice(..) | ty::Adt(..) | ty::Alias(ty::Opaque, _) => {
                 for unsatisfied in unsatisfied_predicates.iter() {
-                    if is_iterator_predicate(unsatisfied.0, self.tcx) {
+                    if is_iterator_predicate(unsatisfied.0) {
                         return true;
                     }
                 }

--- a/tests/ui/iterators/iterator-does-not-need-into-iter.rs
+++ b/tests/ui/iterators/iterator-does-not-need-into-iter.rs
@@ -1,0 +1,18 @@
+//! regression test for #127511: don't suggest `.into_iter()` on iterators
+
+trait Missing {}
+trait HasMethod {
+    fn foo(self);
+}
+impl<T: Iterator + Missing> HasMethod for T {
+    fn foo(self) {}
+}
+
+fn get_iter() -> impl Iterator {
+    core::iter::once(())
+}
+
+fn main() {
+    get_iter().foo();
+    //~^ ERROR the method `foo` exists for opaque type `impl Iterator`, but its trait bounds were not satisfied [E0599]
+}

--- a/tests/ui/iterators/iterator-does-not-need-into-iter.stderr
+++ b/tests/ui/iterators/iterator-does-not-need-into-iter.stderr
@@ -1,0 +1,28 @@
+error[E0599]: the method `foo` exists for opaque type `impl Iterator`, but its trait bounds were not satisfied
+  --> $DIR/iterator-does-not-need-into-iter.rs:16:16
+   |
+LL |     get_iter().foo();
+   |                ^^^ method cannot be called on `impl Iterator` due to unsatisfied trait bounds
+   |
+note: the following trait bounds were not satisfied:
+      `&impl Iterator: Iterator`
+      `&impl Iterator: Missing`
+      `&mut impl Iterator: Missing`
+      `impl Iterator: Missing`
+  --> $DIR/iterator-does-not-need-into-iter.rs:7:9
+   |
+LL | impl<T: Iterator + Missing> HasMethod for T {
+   |         ^^^^^^^^   ^^^^^^^  ---------     -
+   |         |          |
+   |         |          unsatisfied trait bound introduced here
+   |         unsatisfied trait bound introduced here
+   = help: items from traits can only be used if the trait is implemented and in scope
+note: `HasMethod` defines an item `foo`, perhaps you need to implement it
+  --> $DIR/iterator-does-not-need-into-iter.rs:4:1
+   |
+LL | trait HasMethod {
+   | ^^^^^^^^^^^^^^^
+
+error: aborting due to 1 previous error
+
+For more information about this error, try `rustc --explain E0599`.

